### ISSUE TITLE
ENH: all tobii eyetracker data saved

### ISSUE THREAD

**Issue by nmarkowitz** [psychopy/psychopy#4900]
Hi,

I'm using the tobii-spectrum eyetracker. When using the eyetracker with the Psychopy wrapper not all possible data is saved. It looks like the data saved includes:

- gaze position on the monitor
- pupil diameter
- normalized coordinates (track box coordinate system) 

However, the eyetracker returns more data than this. It would be nice to have xyz gaze origin and gaze position as well (user coordinate system). From this you'd be able to get distance from monitor (more accurately and varying over time), visual angle and gaze vector (participant looking at screen from a given angle). Would it be possible to add these? Below are links about these different coordinate systems and data types. 

Thanks for making this really nice wrapper for the eyetracker. It really made it easier to start using it.

Noah

https://developer.tobiipro.com/python/python-sdk-reference-guide.html
https://developer.tobiipro.com/commonconcepts/coordinatesystems.html

**Comment by isolver** [comment]
Could you list the specific tobii sample fields you like added and we can see if they can be added to existing iohub BinocularSample fields not being used by the tobii implementation, or if we will need to add new fields. Thanks very much.

**Comment by nmarkowitz** [comment]
Here's a list of all the keys output from the tobii-spectrum eyetracker:
```
dict_keys(['device_time_stamp', 'system_time_stamp', 'left_gaze_point_on_display_area', 'left_gaze_point_in_user_coordinate_system', 'left_gaze_point_validity', 'left_pupil_diameter', 'left_pupil_validity', 'left_gaze_origin_in_user_coordinate_system', 'left_gaze_origin_in_trackbox_coordinate_system', 'left_gaze_origin_validity', 'right_gaze_point_on_display_area', 'right_gaze_point_in_user_coordinate_system', 'right_gaze_point_validity', 'right_pupil_diameter', 'right_pupil_validity', 'right_gaze_origin_in_user_coordinate_system', 'right_gaze_origin_in_trackbox_coordinate_system', 'right_gaze_origin_validity'])
```

Already recorded by Psychopy are:

* pupil_diamter
* validity
* gaze_point_on_display_area
* gaze_origin_in_trackbox_coordinate_system

It would be nice to also have

* gaze_point_in_user_coordinate_system
* gaze_origin_in_user_coordinate_system

I think gaze_point_on_display_area is converted from 2D to 3D coordinates in psychopy so there's already essentially gaze_point_in_user_coordinate_system. It would be nice to also have the gaze origin in the same coordinate system.

**Comment by peircej** [comment]
Just to clarify, we're talking about the hdf5 file from ioHub? I feel like the PsychoPy csv file saves less than this and the Tobii proprietary file saves everything already, right?

**Comment by isolver** [comment]
Yes, this issue is related to what is saved to the hdf5 file as well as
what is available during the experiment in sample events (which are the
same).

I do not think Tobii trackers have a native data file that it saves
automatically.

**Comment by nmarkowitz** [comment]
There is no native file format for tobii trackers. They rely on you to specify how the data is saved. The hdf5 file saves a key selection of the raw output of from tobii, converts data to same coordinate system psychopy uses and then saves it

**Comment by peircej** [comment]
OK, great, well then it would, indeed be great to add this data entry to our hdf5 outputs. Thanks

**Comment by nmarkowitz** [comment]
If it would help I can send a sample of tobii output with all its fields in Python. It would be a pickled dictionary of the gaze data

**Comment by isolver** [comment]
I think we need to decide if we are going to try and keep a common sample
format across all trackers, move to a more per tracker sample type that
maybe has some shared fields for a subset of the common sample attributes,
or add new fields to the existing sample classes for the important tobii
specific data.

**Comment by isolver** [comment]
Sorry for the delay, I'm looking into addressing this now. To recap, currently iohub uses / saves these fields from the tobii sample:

**tobii field -> iohub fields**

- `gaze_point_on_display_area` (x,y) -> `gaze_x`, `gaze_y`
- `gaze_origin_in_trackbox_coordinate_system` (x,y,z) -> `eye_cam_x`, `eye_cam_y`, `eye_cam_z`

**iohub is not using these fields from tobii samples:**

- `gaze_point_in_user_coordinate_system` (x,y,z)
- `gaze_origin_in_user_coordinate_system` (x,y,z)

@sportnoah14 suggests

> I think gaze_point_on_display_area is converted from 2D to 3D coordinates in psychopy so there's already essentially gaze_point_in_user_coordinate_system. It would be nice to also have the gaze origin in the same coordinate system.

If we do not want change the iohub sample structure, then this would be an option:

1. _If_ current x, y conversion of `gaze_point_on_display_area` to psychopy space is the same as using x, y fields of tobii `gaze_point_in_user_coordinate_system`, then we could populate `gaze_point_in_user_coordinate_system[z]` to the unused iohub `gaze_z` field
2. populate tobii `gaze_origin_in_user_coordinate_system` (x,y,z) to unused iohub fields `angle_x, angle_y, raw_x`

So the iohub sample field _labels_  `angle_x, angle_y, raw_x` would not be meaningful in this case, but the tobii data would be there for users who do want it.

@sportnoah14, what do you think of this idea? For backwards compat, it would be nice if we could keep the iohub sample struct as is imo.
